rightmove_webscraper/__init__.py

```diff
@@ -259,3 +302,190 @@ def _get_results(self, get_floorplans: bool = False):
+        property_data_frame = pd.DataFrame.from_records(final_results)
+
+        def process_data(rawdf):
+            df = rawdf.copy()
+        
+            address = df['address'].str.extract(address_pattern, expand=True).to_numpy()
+            outwardcodes = df['address'].str.extract(outwardcode_pattern, expand=True).to_numpy()
+            
+            df = (df.drop(['address', 'images', 'hasFloorPlan', 'detailUrl'], axis=1)
+                    .assign(address=address[:, 0])
+                    .assign(postcode=address[:, 1])
+                    .assign(outwardcode=outwardcodes[:, 0])
+                    #.assign(transactions=df.transactions.apply(ast.literal_eval))
+                    #.assign(location=df.location.apply(ast.literal_eval))
+                    .assign(last_price=lambda x: extract_price(x.transactions))
+                    .assign(sale_date=lambda x: extract_date(x.transactions))
+                    .assign(tenure=lambda x: extract_tenure(x.transactions))
+                    .assign(lat=lambda x: extract_coords(x.location, lat=True))
+                    .assign(lng=lambda x: extract_coords(x.location))
+                    .drop(['transactions', 'location'], axis=1)
+            )
+            return df
+     
+        #return process_data(property_data_frame)
+
+        return property_data_frame
+
+    @property
+    def processed_data(self):
+        df = self._results
+    
+        address = df['address'].str.extract(address_pattern, expand=True).to_numpy()
+        outwardcodes = df['address'].str.extract(outwardcode_pattern, expand=True).to_numpy()
+        
+        df = (df.drop(['address', 'images', 'hasFloorPlan', 'detailUrl'], axis=1)
+                .assign(address=address[:, 0])
+                .assign(postcode=address[:, 1])
+                .assign(outwardcode=outwardcodes[:, 0])
+                #.assign(transactions=df.transactions.apply(ast.literal_eval))
+                #.assign(location=df.location.apply(ast.literal_eval))
+                .assign(last_price=lambda x: extract_price(x.transactions))
+                .assign(sale_date=lambda x: extract_date(x.transactions))
+                .assign(tenure=lambda x: extract_tenure(x.transactions))
+                .assign(lat=lambda x: extract_coords(x.location, lat=True))
+                .assign(lng=lambda x: extract_coords(x.location))
+                .drop(['transactions', 'location'], axis=1)
+                .reindex(columns=['last_price', 
+                                'sale_date', 
+                                'propertyType',
+                                'bedrooms',
+                                'bathrooms', 
+                                'tenure', 
+                                'address', 
+                                'postcode', 
+                                'outwardcode', 
+                                'lat', 
+                                'lng'])
+        )
+        return df
```

Comment:
Nice one. My only problem is that we recalculate the processed data every time we call this method.

Reply:
Yes, initially I wrote the function inside the ```_get_results``` (code still there, pending decision) but then I thought, maybe one wants to retrieve the raw data anyway ?, ...so I left it as a separate property

in essence, it's a bit like ```.table```, depends on what you think makes more sense? I tailored the processing specifically to how I need to ingest the data (e.g. I only require the last sold price rather than historical info in the transactions data).

I suppose that processing could be more generic and give a tidier df but including all data, if that something that you think is more general purpose?

Reply:
Yes that's perfectly fine.

What I actually meant is that I would have stored the processed data in another data member
`self._processed_data` so that you don't redo the calculation everytime you call `SoldProperties.processed_data`